lib/compat.php

```diff
@@ -464,8 +463,8 @@ function gutenberg_inject_default_block_context( $args ) {
 			}
 
 			// Inject the query context if not done by Core.
-			$needs_query = empty( $block_type->uses_context ) && in_array( 'query', $block_type->uses_context, true );
-			if ( ! isset( $block->context['query'] ) && $needs_query ) {
+			$needs_query = ! empty( $block_type->uses_context ) && in_array( 'query', $block_type->uses_context, true );
+			if ( ! isset( $block->context['query'] ) && $needs_query && ! isset( $block->context['query'] ) ) {
```

Comment:
Not sure if this needs the extra checks you added above. I guess the tests will tell.

Reply:
it doesn't IMO

Reply:
As you can see, the old filter I removed was unsetting just postId and postType. (it was the same code but for the old context implementation) but now that I think about it, I think we need to revert the removal of that filter to solve the issues here..

Reply:
basically we need to revert this removal https://github.com/WordPress/gutenberg/pull/27011/files#diff-9f399722fe93f7131d23a3944e030b88420d66461c7b776e7b153bafd3ae7100L249-L273

Reply:
Ok, I removed my commits. There's still one instance of `in_array` that needs to be guarded for empty values.

Reply:
I added back the filter that was removed and fixed the other instance of `in_array`

Reply:
This check for `isset( $block->context['query'] )` is also redundant (it's already earlier in the same line), so I removed that too